KEYCLOAK-10983 Namespace support for LDAP group mapper

tjuerge opened the pull request:
Previously, Keycloak did only support syncing groups from LDAP federation provider as top-level KC groups.

This approach has some limitations:
- If using multiple group mappers then there’s no way to isolate the KC groups synched by each group mapper.
- If the option "Drop non-existing groups during sync” is activated then all KC groups (including the manually created ones) are deleted.
- There’s no way to inherit roles from a parent KC group.

This patch introduces support to specify a prefix for the resulting group path, which effectively serves as a namespace for a group.

A path prefix can be specified via the newly introduced `Groups Path` config option on the mapper. This groups path defaults to `/` for top-level groups.

This also enables to have multiple `group-ldap-mapper`'s which can manage groups within their own namespace.

An `group-ldap-mapper` with a `Group Path` configured as `/Applications/App1` will only manage groups under that path. Other groups, either manually created or managed by other `group-ldap-mapper` are not affected.

<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapperFactory.java:
```diff
@@ -198,6 +198,12 @@
+                .add()
+                .property().name(GroupMapperConfig.LDAP_GROUPS_PATH)
+                .label("Groups Path")
+                .helpText("Keycloak group path the LDAP groups are added to")
```
Is it possible to change tooltip to contain more details? I would use something like: _Keycloak group path the LDAP groups are added to. For example if value '/Applications/App1' is used, then LDAP groups will be available in Keycloak under group 'App1', which is child of top level group 'Applications'. The default value is '/' so LDAP groups will be mapped to the Keycloak groups at the top level. The configured group path must already exists in the Keycloak when creating this mapper._

tjuerge replied:
Sure. If you don't mind then I'll use your tooltip text as-is 👍

mposolda reviewed (changes requested):
@tjuerge Thanks for the PR. It is great work.

I've added one inline comment regarding more details to the tooltip, so that it is more obvious what is supposed to be used for this configuration value.

Besides that, I suggest to require that configured Keycloak group must already exists. This will allow that all the changes in KeycloakModelUtils can be removed (the changes for lazy group creation are quite complex, so removing that will greatly simplify this PR). Also you mention that often there is a need for example to somehow configure the groups (EG. you usually want to add some roles under top level '/Applications' before you configure the group '/Applications' as a namespace for this mapper). The only needed thing may be to add something to GroupLDAPStorageMapperFactory.validateConfiguration and make sure that Keycloak group already exists - this should be quite easy due the already existing method KeycloakModelUtils.findGroupByPath .

tjuerge commented:
> Besides that, I suggest to require that configured Keycloak group must already exists.

@mposolda Thanks for your suggestion and the hint regarding `GroupLDAPStorageMapperFactory.validateConfiguration()`. I wasn't aware of the validation support. This approach is definitely better then automagically creating missing nested groups under the hood.

I'll come up with the corresponding implementation.

mposolda commented:
@tjuerge Thanks!

tjuerge commented:
@mposolda The tooltip text has been updated and the lazy group creation has been removed.
`GroupLDAPStorageMapperFactory.validateConfiguration()` has a check for the groups path group now.

tjuerge commented:
I closed this PR by accident.

mposolda commented on federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/membership/group/GroupLDAPStorageMapperFactory.java:
```diff
@@ -282,6 +292,12 @@ public void validateConfiguration(KeycloakSession session, RealmModel realm, Com
+        checkMandatoryConfigAttribute(GroupMapperConfig.LDAP_GROUPS_PATH, "Groups Path", config);
+        String group = config.getConfig().getFirst(GroupMapperConfig.LDAP_GROUPS_PATH).trim();
+        if (!"/".equals(group) && KeycloakModelUtils.findGroupByPath(realm, group) == null) {
+            throw new ComponentValidationException("Groups path group '" + group + "' does not exist - please create it first");
```
Is it possible to use generic message without adding "group" variable into the text? Maybe just something like "Groups path group does not exist  - please create the group on specified path first" .

The reason is some potential for various injection attacks, which would be possible if the user input (variable "group") is shown in the notification in the UI.

Bonus point: If you manage to localize the error message, it will be nice as well. The message with specified key will just need to be added to this file and that should be enough: https://github.com/keycloak/keycloak/blob/master/themes/src/main/resources/theme/base/admin/messages/messages_en.properties

tjuerge replied:
Sure. And now I know why the other exception messages are so generic.

tjuerge replied:
And here we go

mposolda reviewed (changes requested):
@tjuerge Thanks for the update! I've added one inline comment, but I hope it should be just small detail to update and we will be able to merge this :-)

tjuerge committed:
Namespace support to group-ldap-mapper

Previously, Keycloak did only support syncing groups from LDAP federation provider as top-level KC groups.

This approach has some limitations:
- If using multiple group mappers then there’s no way to isolate the KC groups synched by each group mapper.
- If the option "Drop non-existing groups during sync” is activated then all KC groups (including the manually created ones) are deleted.
- There’s no way to inherit roles from a parent KC group.

This patch introduces support to specify a prefix for the resulting group path, which effectively serves as a namespace for a group.

A path prefix can be specified via the newly introduced `Groups Path` config option on the mapper. This groups path defaults to `/` for top-level groups.

This also enables to have multiple `group-ldap-mapper`'s which can manage groups within their own namespace.

An `group-ldap-mapper` with a `Group Path` configured as `/Applications/App1` will only manage groups under that path. Other groups, either manually created or managed by other `group-ldap-mapper` are not affected.

mposolda reviewed (approved):
@tjuerge That was quick :) Thanks!

tjuerge commented:
@mposolda Thanks for your support..

dimitry-a-baranov commented:
Hi everyone!

I have a KC 15.0.2 setup with the following configuration:
Single LDAP Mapper in WRITABLE mode targeted to ActiveDirectory Domain with multiple OU's where KC-managed groups are contained.
Several (e.g. 4) ldap-group-mapper's with LDAP_ONLY mode configured to different Groups Path's finally resolving groups in two-level hierarchy:
/ (ldap-group-mapper-0 with LDAP Groups DN 0)
/Groups-Path1/ (ldap-group-mapper-1 with LDAP Groups DN 1)
/Groups-Path2/ (ldap-group-mapper-2 with LDAP Groups DN 2)
/Groups-Path3/ (ldap-group-mapper-3 with LDAP Groups DN 3)
On sync'ing the groups from LDAP everything works fine, the groups from subsequent LDAP Groups DN's are successfully imported into configured group paths, but when adding the user in KC GUI to some of the imported groups fails on ActiveDirectory side with error UpdErr: DSID-031A11DA, problem 6005 (ENTRY_EXISTS), data 0 meaning that KC is trying to create or update a group with existing name (cn) in an incorrect OU of the domain and ActiveDirectory does not allow that since group names must be unique throughout the domain. In other words, despite the specifically attributed Group Path the KC is selecting wrong group mapper for updating the group attributes/membership and ActiveDirectory is refusing the update to the group structure as soon as DN is wrong.
This error appears with every group mapper groups except this, related to the "automatically" selected wrong mapper. For the wrong mapper, group attributes/membership management proceeds without any problems.
Another symptom is that with the mentioned setup the current user groups in KC user groups GUI section are shown only for that mapper, while users are shown correctly if listed through KC group members GUI section (therefore the same listings vary in two sections of the GUI).
Please help to resolve this problem or route me to solution if already resolved.
